File: src/bmsx/core/service.ts
import { type RegisterablePersistent, type Identifier, type Identifiable } from '../rompack/rompack';
import { EventEmitter, EventPort, eventsOf } from './eventemitter';
import { Stateful } from '../fsm/fsmtypes';
import { StateMachineController } from '../fsm/fsmcontroller';
import { StateDefinitionBuilders } from '../fsm/fsmdecorators';
import { Registry } from './registry';
import { onload, excludepropfromsavegame, type RevivableObjectArgs } from '../serializer/serializationhooks';

function deriveIdFromConstructor(name: string): Identifier {
	// Convert PascalCase/MyService to lower_snake_case: my_service
	const snake = name
		.replace(/^_+/, '') // strip leading underscores
		.replace(/([a-z0-9])([A-Z])/g, '$1_$2') // aA -> a_A
		.replace(/[-\s]+/g, '_') // dashes/spaces -> underscore
		.replace(/__+/g, '_') // collapse duplicate underscores
		.toLowerCase();
	return (snake || 'service') as Identifier;
}

/**
 * Base class for non-world-bound, persistent services (UE-style Subsystems).
 *
 * Characteristics:
 * - Registers in the global Registry at construction time.
 * - Default `registrypersistent = true` so services survive world loads unless explicitly disposed.
 * - Opt-in event processing is governed by `eventhandling_enabled` (true by default).
 * - Decorator-driven event subscriptions are initialized via the shared decorator logic
 *   (they will run after construction and after Registry.register due to deferred init).
 */
export class Service implements Stateful, Identifiable, RegisterablePersistent {
	/** Unique identifier for the service (override or pass via constructor). */
	public id: Identifier;
	@excludepropfromsavegame
	public readonly events: EventPort;
	sc: StateMachineController;
	/** True when the service participates in gameplay. */
	public active: boolean = false;
	/** If false, systems should not advance time-based logic for this service. */
	public tickEnabled: boolean = true;

	/**
	 * If true, service survives Registry.clear() calls (e.g., during world reloads).
	 */
	public get registrypersistent(): true { return true; }

	/**
	 * Controls whether the service processes events (checked by EventEmitter gating).
	 * Services default to listening immediately; subclasses may toggle as needed.
	 */
	public eventhandling_enabled: boolean = false;

	// Default no-op state hooks; subclasses override to participate in Savegame.
	// TODO: WILL THIS WORK WITH LUA EXTENSIONS AND THE LUA RUNTIME WHERE ONLY BASE SERVICE IS CREATED?
	public getState(): unknown { return undefined; }
	public setState(_: unknown): void { /* no-op by default */ }

	/**
	 * Construct a new Service.
	 * @param id Unique identifier. If omitted, defaults to the class name in lower_snake_case.
	 * @param opts Optional flags for initial state. Pass `deferBind: true` to delay registration until the subclass calls `bind()`.
	 */
	public constructor(opts?: RevivableObjectArgs & { id?: Identifier; deferBind?: boolean }) {
		this.id = opts?.id ?? deriveIdFromConstructor(this.constructor.name ?? 'service');
		this.events = eventsOf(this);

		const fsmName = this.constructor.name;
		const hasDef = !!StateDefinitionBuilders[fsmName];
		this.sc = hasDef ? new StateMachineController({ ...opts, fsm_id: fsmName, id: this.id }) : new StateMachineController(opts);

		// Register service in global registry
		if (opts?.deferBind !== true) {
			this.bind();
		}
	}

	/**
	 * Convenience: enable event processing for this service.
	 */
	public enableEvents(): void { this.eventhandling_enabled = true; }

	/**
	 * Convenience: disable event processing for this service.
	 */
	public disableEvents(): void { this.eventhandling_enabled = false; }

	/**
	 * Dispose the service; unsubscribes from events and deregisters from the Registry.
	 * Note: persistent records are preserved by default in this codebase; pass `true` to
	 * Registry.deregister's second argument if removal of the persistent record is required.
	 */
	public dispose(): void {
		// Remove all event subscriptions for this service and mark it inactive.
		this.disableEvents();
		// Deregister, preserving persistence behavior unless caller forces removal.
		this.unbind();
	}

	/** Wire decorator-declared subscriptions for this service. */
	@onload
	public bind(): void {
		Registry.instance.register(this);
		// Bind controller subscriptions (no start on revive/bind)
		if (!this.sc) {
			throw new Error(`[Service:${this.id}] State machine controller was not initialized before bind().`);
		}
		this.sc.bind();
	}

	/** Unwire all subscriptions for this service. */
	public unbind(): void {
		EventEmitter.instance.removeSubscriber(this);
		Registry.instance.deregister(this, true);
	}

	/** BeginPlay-style activation for services (starts FSM). */
	public activate(): void {
		this.active = true;
		this.bind();
		this.enableEvents();
		this.sc.start();
	}

	public deactivate(): void {
		this.active = false;
		this.disableEvents();
		this.sc.pause();
	}
}

File: src/bmsx/res/bios/service.lua
-- service.lua
-- lightweight service base with fsm hook

local eventemitter = require("eventemitter")
local fsm = require("fsm")
local fsmlibrary = require("fsmlibrary")

local registry = require("registry")

local service = {}
service.__index = service

function service.new(opts)
	local self = setmetatable({}, service)
	opts = opts or {}
	self.id = opts.id or "service"
	self.type_name = "service"
	self.registrypersistent = opts.registrypersistent ~= false
	self.active = false
	self.tick_enabled = true
	self.eventhandling_enabled = false
	self.events = eventemitter.events_of(self)
	local definition = opts.definition or (opts.fsm_id and fsmlibrary.get(opts.fsm_id))
	self.sc = opts.sc or fsm.statemachinecontroller.new({ target = self, definition = definition, fsm_id = opts.fsm_id })
	return self
end

function service:enable_events()
	self.eventhandling_enabled = true
end

function service:disable_events()
	self.eventhandling_enabled = false
end

function service:activate()
	self.active = true
	self:enable_events()
	self.sc:start()
	self.sc:resume()
end

function service:deactivate()
	self.active = false
	self:disable_events()
	self.sc:pause()
end

-- Forwards an event to the FSM, which can be used for state transitions or as a general-purpose event bus for the service's internal logic. The event can be a string (event type) or a table (event object). If it's a string, it will be wrapped in a table with the type and emitter fields.
function service:dispatch_state_event(event_or_name, payload)
	return self.sc:dispatch(event_or_name, payload)
end

-- Useless alias for dispatch_state_event, but provided for semantic clarity in some cases, so that Codex can recognize the intent as dispatching a command rather than a gameplay fact.
function service:dispatch_command(event_or_name, payload)
	return self.sc:dispatch(event_or_name, payload)
end

function service:dispose()
	self:disable_events()
	eventemitter.eventemitter.instance:remove_subscriber(self)
	self.sc:dispose()
	registry.instance:deregister(self, true)
end

return service
